chapter.md

```diff
@@ -0,0 +1,735 @@
+then subscribe to that with a another fat arrow callback
+that updates the form data
+(which as explained above triggers redisplay).
+
+The second half of `ngOnInit` does the binding in the opposite direction:
+whenever the form changes,
+we dispatch an action created by `saveForm`
+that has `SAVE_FORM` as the change
+and `['character']` as the path to the part of the state we want to modify.
+FIXME: how does `SAVE_FORM` get into this as `change`?
+
+And that's it:
+every change to our form triggers creation of a new state,
+and every change to our state triggered by anything else
+is immediately reflected in our form.
+We *aren't* saving the state to disk,
+but that's best left to some kind of middleware
+(which can be smart about only persisting the bits of state
+that have actually changed between updates).
+
+## Filling in the Gaps
+
+The only meaningful way to assess an architectural decision is to ask
+whether it makes things comprehensible today
+and easy to change tomorrow.
+Given all of the plumbing introduced above,
+you might think that our architecture falls short on both counts,
+but it turns out that it actually simplifies application evolution.
+To see this,
+let's have a look at how it helps us handle multi-valued forms,
+which Angular doesn't support out of the box.
+
+As the name suggests,
+a multi-valued form is one that can have many values at once,
+such as a multi-select dropdown list.
+The natural way to store these values is in an array,
+such as the one shown below for storing a character's skills:
+
+```ts
+character: {
+  skills: ['Drinking', 'Knowing Things'],
+}
+```
+
+The first step in adding support for multi-valued fields to our application
+is to define actions that manipulate arrays–or more precisely,
+to define functions that create actions that tell Redux
+how to generate a new array from an old one:
+
+```ts
+const addIntoArray = (path, value) => ({
+  type: 'ADD_INDEXED_FORM_VALUE',
+  payload: { path, value }
+});
+
+const putInArray = (path, value, index) => ({
+  type: 'UPDATE_INDEXED_FORM_VALUE',
+  payload: { path, value, index }
+});
+
+const removeFromArray = (index, path) => ({
+  type: 'REMOVE_INDEXED_FORM_VALUE',
+  payload: { index, path }
+});
+```
+
+There's nothing exciting going on here:
+each function creates an action with a `type` (again, required by Redux)
+and a `payload` (our own term).
+To cut down on typing,
+we make use of the fact that `{a, b}` is a shorthand for
+`{a: a, b: b}`:
+
+```js
+let a = 'A';
+let b = 'B';
+let both = {a, b};
+console.log(both);
+```
+```
+{a: 'A', b: 'B'}
+```
+
+The payload of the action returned by `addIntoArray`
+therefore has two keys `path` and `value`,
+which are in turn bound to whatever was passed in
+for the parameters with the corresponding names.
+
+Once we have these actions,
+the next step is to update our reducer by adding the following case:
+
+```ts
+case 'ADD_INDEXED_FORM_VALUE':
+  const lensForProp = lensPath(action.payload.path);
```

Comment:
this all can be simplified with the `path`

^ sorry about that, I have to update the source